### PR TITLE
Skip execution and delete the changeset if it has changes

### PR DESCRIPTION
**Pull request by pflugs30**
## Description
This PR modifies the **AWS CloudFormation Execute Change Set** to skip the execution of a given change set if it has no changes and is in a FAILED state.  The task will also delete the empty change set and finish the task operation with success, allowing a pipeline to continue.

## Motivation
The **AWS CloudFormation Execute Change Set** task will fail if the change set to execute is in a FAILED state due to having no changes.  This can be a common occurrence in some pipelines with a separate stage for build and deploy of CloudFormation stacks.  Without having logic to identify an empty change set, the pipeline itself will need to have logic to determine the empty set and handle the condition.

## Related Issue(s), If Filed
See #142.

## Testing
I've added a unit test to `cloudFormationExecuteChangeSet-test.ts` to cover the additional logic.  I will also be installing the extension on my client's corporate Azure DevOps hosted environment and testing in an organization with several pipelines.

## Checklist
-   [X] I have read the **README** document
-   [X] I have read the **CONTRIBUTING** document
-   [X] My code follows the code style of this project
-   [X] I have added tests to cover my changes
-   [X] A short description of the change has been added to the changelog using the script `npm run newChange`

## License
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by pflugs30** [comment]
Just a quick thought... if desired by project admins or others, I'd be willing to add a switch to toggle the deletion of the empty change set.  I can't imagine why someone would want to leave it around, but hey, it takes all kinds. :-)

File: src/tasks/CloudFormationExecuteChangeSet/TaskOperations.ts [hunk 1]
```diff
@@ -20,29 +20,50 @@ export class TaskOperations {
...
-            await this.cloudFormationClient
-                .executeChangeSet({
-                    ChangeSetName: this.taskParameters.changeSetName,
-                    StackName: this.taskParameters.stackName
-                })
-                .promise()
+            const changeSetHasNoChanges =
+                changeSet.Status === 'FAILED' &&
+                changeSet.StatusReason ===
+                    "The submitted information didn't contain changes. Submit different information to create a change set."
```
Comment (bryceitoc9) [inline src/tasks/CloudFormationExecuteChangeSet/TaskOperations.ts:37]: See thoughts in the main comment but at the very least we should tie this message (and the message in the test) to match https://github.com/aws/aws-toolkit-azure-devops/blob/62d2a6a786db05c5ea4ccae1e205713a8e89c1bc/src/tasks/CloudFormationCreateOrUpdateStack/TaskOperations.ts#L532 ; the CLI is also testing the same strings: https://github.com/aws/aws-cli/blob/45b0063b2d0b245b17a57fd9eebd9fcc87c4426a/awscli/customizations/cloudformation/deployer.py#L171-L174

**Review by bryceitoc9 — COMMENTED** [review]
Thanks for the contribution, @pflugs30! I've got a few (slightly disorganized) thoughts on this:

* Notably, we are already doing this check in the Update flow: https://github.com/aws/aws-toolkit-azure-devops/blob/62d2a6a786db05c5ea4ccae1e205713a8e89c1bc/src/tasks/CloudFormationCreateOrUpdateStack/TaskOperations.ts#L532 ; I'm wondering if it would be a better idea to do it there and pass along a variable to make this a no-op on execution.
* I'm thinking that it's a good idea at this point to include the toggle since it appears no other AWS constructs automatically delete the changeset (e.g. https://github.com/aws/aws-cli/issues/4534 ) , and I don't want us to have a "surprising" behavior included by default.
* I'm thinking of changing this to output a "SucceededWithIssues" or a "Skipped" result in these cases, although that would involve modifying our `generateRunners` code (I don't expect you to take that on for this PR). Would that make sense to you as a user to signal that the change set was not "executed"?

**Comment by pflugs30** [comment]
> * Notably, we are already doing this check in the Update flow: https://github.com/aws/aws-toolkit-azure-devops/blob/62d2a6a786db05c5ea4ccae1e205713a8e89c1bc/src/tasks/CloudFormationCreateOrUpdateStack/TaskOperations.ts#L532
> I'm wondering if it would be a better idea to do it there and pass along a variable to make this a no-op on execution.

I'm not sure what you mean by "do it there".  Do you mean use the empty changeset check there?  Now that I know about this other check, I'd be more likely to refactor it to a utility function in the `lib` folder or something.

> * I'm thinking that it's a good idea at this point to include the toggle since it appears no other AWS constructs automatically delete the changeset (e.g. [aws/aws-cli#4534](https://github.com/aws/aws-cli/issues/4534) ) , and I don't want us to have a "surprising" behavior included by default.

I agree, especially after seeing the issue you referenced above.  I will add a toggle.

> * I'm thinking of changing this to output a "SucceededWithIssues" or a "Skipped" result in these cases, although that would involve modifying our `generateRunners` code (I don't expect you to take that on for this PR). Would that make sense to you as a user to signal that the change set was not "executed"?

I like having a specific status, and I'd prefer to see something like "Skipped".

File: src/tasks/CloudFormationExecuteChangeSet/TaskOperations.ts [hunk 2]
```diff
@@ -20,29 +20,50 @@ export class TaskOperations {
...
+
+            if (changeSetHasNoChanges) {
```
[the remark this answers is not in the document]
Comment (pflugs30) [reply src/tasks/CloudFormationExecuteChangeSet/TaskOperations.ts]: Yes, I think it still makes sense to perform these operations given that the stack will still exist (and could have Output values) even if the change set is deleted.  In my use cases, I will have several related stacks in the same pipeline, and I can foresee instances where I'd want access to the outputs in the pipelines.

File: src/tasks/CloudFormationExecuteChangeSet/TaskOperations.ts (hunk 1 above)
Comment (pflugs30) [reply src/tasks/CloudFormationExecuteChangeSet/TaskOperations.ts]: I've consolidated these checks into a utility function.

**Comment by pflugs30** [comment]
@bryceitoc9 I've committed several new changes that implement your suggestions and return the task's default behavior to the pre-PR state.  Please review at your convenience.

**Comment by bryceitoc9** [comment]
> I'm not sure what you mean by "do it there". Do you mean use the empty changeset check there? Now that I know about this other check, I'd be more likely to refactor it to a utility function in the lib folder or something.

We're already doing an empty changeset check there in the create/update task, so it might be worth leaving the check there and triggering that to skip the whole execute step (e.g. returning an empty variable). I think that burden might be put onto the person constructing the pipeline, though, and it would probably get ambiguous if someone were working with more than one stack in a single pipeline flow, so it's probably best to have the check here (unless you think there's a valid case to create/update in a pipeline without ever executing, even if the changeset isn't empty)

> I like having a specific status, and I'd prefer to see something like "Skipped".

I'll take a look into this, I might still have some other state (maybe "Success") in the case of deletion since that's indicating that the step is still doing something. Thanks for the feedback!

File: src/lib/cloudformationutils.ts [hunk 3]
```diff
@@ -148,3 +148,29 @@ export async function testChangeSetExists(
...
+        /^No updates are to be performed./,
+        /^The submitted information didn't contain changes./
+    ]
+
+    errCodeOrStatus = errCodeOrStatus || ''
+    const message = errMessage || ''
+    if (errCodeOrStatus.search(/ValidationError/) !== -1 || errCodeOrStatus.search(/FAILED/) !== -1) {
+        knownNoOpErrorMessages.forEach(element => {
+            if (message.search(element) !== -1) {
+                noWorkToDo = true
```
Comment (bryceitoc9) [inline src/lib/cloudformationutils.ts:170]: May as well just `return true` at this point, no point in continuing the forEach. That way, you don't need the `let` and can also just `return false` at the end.

File: src/tasks/CloudFormationExecuteChangeSet/task.json [hunk 4]
```diff
@@ -60,6 +60,22 @@
             "required": false,
             "helpMarkDown": "The stack name or Amazon Resource Name (ARN) of the stack associated with the change set. This value is required if you specified the name of a change set to execute. If the ARN of the change set was specified this field is optional.\n\nThe name must be unique in the region in which you are creating the stack. A stack name can contain only alphanumeric characters (case-sensitive) and hyphens. It must start with an alphabetic character and cannot be longer than 128 characters."
         },
+        {
+            "name": "noFailOnEmptyChangeSet",
+            "type": "boolean",
+            "label": "Do Not Fail On Empty Change Set",
+            "defaultValue": "false",
+            "required": false,
+            "helpMarkDown": "Causes the task to identify an empty change set and skip its excution without throwing an error."
```
Comment (bryceitoc9) [inline src/tasks/CloudFormationExecuteChangeSet/task.json:69]: Nit: spelling

```suggestion
            "helpMarkDown": "Causes the task to identify an empty change set and skip its execution without throwing an error."
```

File: src/tasks/CloudFormationExecuteChangeSet/task.json [hunk 5]
```diff
@@ -60,6 +60,22 @@
...
+        },
+        {
+            "name": "deleteEmptyChangeSet",
+            "type": "boolean",
+            "label": "Delete Empty Change Set",
+            "defaultValue": "false",
+            "required": false,
+            "helpMarkDown": "Toggles the deletion of an empty change set.  Only used if `noFailOnEmptyChangeSet=true`."
```
Comment (bryceitoc9) [inline src/tasks/CloudFormationExecuteChangeSet/task.json:77]: Adding a `"visibleRule": "noFailOnEmptyChangeSet = true"` entry should cover this case, in which case you should also remove the note here. Help note can probably read "Automatically delete change set if empty."

**Review by bryceitoc9 — COMMENTED** [review]
Mainly nits at this point. Looking good, I'll give it a manual test before approval. Thanks for the quick turnaround!

File: src/lib/cloudformationutils.ts (hunk 3 above)
Comment (pflugs30) [reply src/lib/cloudformationutils.ts:170]: Done

File: src/tasks/CloudFormationExecuteChangeSet/task.json (hunk 5 above)
Comment (pflugs30) [reply src/tasks/CloudFormationExecuteChangeSet/task.json:77]: Done

**Comment by pflugs30** [comment]
I convinced our Azure DevOps owner to install the test package I built for local testing.  I was able to execute my pipeline, and it correctly skipped and deleted the empty change sets.  Here's the log for reference:

```
2021-11-19T00:36:55.9389368Z ##[section]Starting: Deploy Changesets
2021-11-19T00:36:56.1353373Z ##[section]Starting: Initialize job
2021-11-19T00:36:56.1354733Z Agent name: 'Hosted Agent'
2021-11-19T00:36:56.1355179Z Agent machine name: 'fv-az48-24'
2021-11-19T00:36:56.1355410Z Current agent version: '2.195.0'
2021-11-19T00:36:56.1397204Z ##[group]Operating System
2021-11-19T00:36:56.1397474Z Ubuntu
2021-11-19T00:36:56.1397616Z 20.04.3
2021-11-19T00:36:56.1397766Z LTS
2021-11-19T00:36:56.1397913Z ##[endgroup]
2021-11-19T00:36:56.1398087Z ##[group]Virtual Environment
2021-11-19T00:36:56.1398308Z Environment: ubuntu-20.04
2021-11-19T00:36:56.1398498Z Version: 20211114.1
2021-11-19T00:36:56.1398823Z Included Software: https://github.com/actions/virtual-environments/blob/ubuntu20/20211114.1/images/linux/Ubuntu2004-README.md
2021-11-19T00:36:56.1399225Z Image Release: https://github.com/actions/virtual-environments/releases/tag/ubuntu20%2F20211114.1
2021-11-19T00:36:56.1399500Z ##[endgroup]
2021-11-19T00:36:56.1399688Z ##[group]Virtual Environment Provisioner
2021-11-19T00:36:56.1400066Z 1.0.0.0-master-20211108-1
2021-11-19T00:36:56.1400240Z ##[endgroup]
2021-11-19T00:36:56.1401136Z Current image version: '20211114.1'
2021-11-19T00:36:56.1403641Z Agent running as: 'vsts'
2021-11-19T00:36:56.1444304Z Prepare build directory.
2021-11-19T00:36:56.1695535Z Set build variables.
2021-11-19T00:36:56.1726026Z Download all required tasks.
2021-11-19T00:36:56.1833097Z Downloading task: CloudFormationExecuteChangeSet (1.13.0)
2021-11-19T00:36:57.0718308Z Checking job knob settings.
2021-11-19T00:36:57.0726989Z    Knob: AgentToolsDirectory = /opt/hostedtoolcache Source: ${AGENT_TOOLSDIRECTORY} 
2021-11-19T00:36:57.0728176Z    Knob: AgentPerflog = /home/vsts/perflog Source: ${VSTS_AGENT_PERFLOG} 
2021-11-19T00:36:57.0729270Z Finished checking job knob settings.
2021-11-19T00:36:57.0973643Z Start tracking orphan processes.
2021-11-19T00:36:57.1162445Z ##[section]Finishing: Initialize job
2021-11-19T00:36:57.1423841Z ##[section]Starting: Deploy Network stack changes
2021-11-19T00:36:57.1658885Z ==============================================================================
2021-11-19T00:36:57.1659554Z Task         : AWS CloudFormation Execute Change Set
2021-11-19T00:36:57.1660031Z Description  : Executes a change set to create or update a stack.
2021-11-19T00:36:57.1660287Z Version      : 1.13.0
2021-11-19T00:36:57.1660665Z Author       : Amazon Web Services
2021-11-19T00:36:57.1662113Z Help         : Please refer to [AWS CloudFormation User Guide](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/) for more information on working with AWS CloudFormation.

More information on this task can be found in the [task reference](https://docs.aws.amazon.com/vsts/latest/userguide/cloudformation-execute-changeset.html).

####Task Permissions
This task requires permissions to call the following AWS service APIs (depending on selected task options, not all APIs may be used):
* cloudformation:DescribeStacks
* cloudformation:DescribeChangeSet
* cloudformation:DescribeStackResources
* cloudformation:ExecuteChangeSet
2021-11-19T00:36:57.1663270Z ==============================================================================
2021-11-19T00:36:57.1789352Z ##[warning]This task uses Node 6 execution handler, which will be deprecated soon. If you are the developer of the task - please consider the migration guideline to Node 10 handler - https://aka.ms/migrateTaskNode10. If you are the user - feel free to reach out to the owners of this task to proceed on migration.
2021-11-19T00:36:57.8407748Z Configuring credentials for task
2021-11-19T00:36:57.8411540Z ...configuring AWS credentials from service endpoint '1ceb425b-70d2-43df-b248-da2eb6957310'
2021-11-19T00:36:57.8412624Z ...endpoint defines standard access/secret key credentials
2021-11-19T00:36:57.8424463Z Configuring region for task
2021-11-19T00:36:57.8426511Z ...configured to use region us-east-1, defined in task.
2021-11-19T00:36:58.4145648Z Skipping execution of change set network-changeset since it has no pending changes
2021-11-19T00:36:58.4147621Z Deleting change set network-changeset since it has no pending changes
2021-11-19T00:36:58.4148312Z Executed change set network-changeset
2021-11-19T00:36:58.4194984Z ##[section]Finishing: Deploy Network stack changes
2021-11-19T00:36:58.4223237Z ##[section]Starting: Deploy EB Apps stack changes
2021-11-19T00:36:58.4232596Z ==============================================================================
2021-11-19T00:36:58.4232931Z Task         : AWS CloudFormation Execute Change Set
2021-11-19T00:36:58.4233288Z Description  : Executes a change set to create or update a stack.
2021-11-19T00:36:58.4233528Z Version      : 1.13.0
2021-11-19T00:36:58.4233807Z Author       : Amazon Web Services
2021-11-19T00:36:58.4234992Z Help         : Please refer to [AWS CloudFormation User Guide](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/) for more information on working with AWS CloudFormation.

More information on this task can be found in the [task reference](https://docs.aws.amazon.com/vsts/latest/userguide/cloudformation-execute-changeset.html).

####Task Permissions
This task requires permissions to call the following AWS service APIs (depending on selected task options, not all APIs may be used):
* cloudformation:DescribeStacks
* cloudformation:DescribeChangeSet
* cloudformation:DescribeStackResources
* cloudformation:ExecuteChangeSet
2021-11-19T00:36:58.4236190Z ==============================================================================
2021-11-19T00:36:58.4369559Z ##[warning]This task uses Node 6 execution handler, which will be deprecated soon. If you are the developer of the task - please consider the migration guideline to Node 10 handler - https://aka.ms/migrateTaskNode10. If you are the user - feel free to reach out to the owners of this task to proceed on migration.
2021-11-19T00:37:00.0695659Z Configuring credentials for task
2021-11-19T00:37:00.0697508Z ...configuring AWS credentials from service endpoint '1ceb425b-70d2-43df-b248-da2eb6957310'
2021-11-19T00:37:00.0697997Z ...endpoint defines standard access/secret key credentials
2021-11-19T00:37:00.0698376Z Configuring region for task
2021-11-19T00:37:00.0699467Z ...configured to use region us-east-1, defined in task.
2021-11-19T00:37:00.0701201Z Skipping execution of change set eb-apps-changeset since it has no pending changes
2021-11-19T00:37:00.0702492Z Deleting change set eb-apps-changeset since it has no pending changes
2021-11-19T00:37:00.0702997Z Executed change set eb-apps-changeset
2021-11-19T00:37:00.0707138Z ##[section]Finishing: Deploy EB Apps stack changes
2021-11-19T00:37:00.0726298Z ##[section]Starting: Deploy Create AMI Shared resources stack changes
2021-11-19T00:37:00.0733356Z ==============================================================================
2021-11-19T00:37:00.0733673Z Task         : AWS CloudFormation Execute Change Set
2021-11-19T00:37:00.0733971Z Description  : Executes a change set to create or update a stack.
2021-11-19T00:37:00.0734210Z Version      : 1.13.0
2021-11-19T00:37:00.0734454Z Author       : Amazon Web Services
2021-11-19T00:37:00.0735548Z Help         : Please refer to [AWS CloudFormation User Guide](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/) for more information on working with AWS CloudFormation.

More information on this task can be found in the [task reference](https://docs.aws.amazon.com/vsts/latest/userguide/cloudformation-execute-changeset.html).

####Task Permissions
This task requires permissions to call the following AWS service APIs (depending on selected task options, not all APIs may be used):
* cloudformation:DescribeStacks
* cloudformation:DescribeChangeSet
* cloudformation:DescribeStackResources
* cloudformation:ExecuteChangeSet
2021-11-19T00:37:00.0736672Z ==============================================================================
2021-11-19T00:37:00.0777949Z ##[warning]This task uses Node 6 execution handler, which will be deprecated soon. If you are the developer of the task - please consider the migration guideline to Node 10 handler - https://aka.ms/migrateTaskNode10. If you are the user - feel free to reach out to the owners of this task to proceed on migration.
2021-11-19T00:37:01.0026361Z Configuring credentials for task
2021-11-19T00:37:01.0028159Z ...configuring AWS credentials from service endpoint '1ceb425b-70d2-43df-b248-da2eb6957310'
2021-11-19T00:37:01.0028637Z ...endpoint defines standard access/secret key credentials
2021-11-19T00:37:01.0029021Z Configuring region for task
2021-11-19T00:37:01.0030110Z ...configured to use region us-east-1, defined in task.
2021-11-19T00:37:01.0031914Z Skipping execution of change set create-ami-shared-changeset since it has no pending changes
2021-11-19T00:37:01.0033226Z Deleting change set create-ami-shared-changeset since it has no pending changes
2021-11-19T00:37:01.0033942Z Executed change set create-ami-shared-changeset
2021-11-19T00:37:01.0037496Z ##[section]Finishing: Deploy Create AMI Shared resources stack changes
2021-11-19T00:37:01.0089665Z ##[section]Starting: Finalize Job
2021-11-19T00:37:01.0118842Z Cleaning up task key
2021-11-19T00:37:01.0120288Z Start cleaning up orphan processes.
2021-11-19T00:37:01.0391222Z ##[section]Finishing: Finalize Job
2021-11-19T00:37:01.0430721Z ##[section]Finishing: Deploy Changesets
```

**Comment by bryceitoc9** [comment]
Nice, running a couple tests on our side but it's looking good so far,.

**Review by bryceitoc9 — APPROVED** [review]
Apologies about the delay, but things are looking good on my end after some manual testing! Thanks for the contribution!

**Comment by pflugs30** [comment]
Thank you, @bryceitoc9 !

**Pull request merged**